psdisk/psdisk.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Phoenix Systems disk tool
+ *
+ * Copyright 2020 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <stdio.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include <sys/list.h>
+
+/* TODO: change access to ptable.h */
+#include "../../phoenix-rtos-filesystems/ptable/ptable.h"
+
+
+#define LOG_ERROR(str, ...) do { fprintf(stderr, __FILE__  ":%d error: " str "\n", __LINE__, ##__VA_ARGS__); } while (0)
+
+
+/* TODO: create partitions list */
+struct {
+	uint32_t pCnt;
+	memory_properties_t mem;
+
+	FILE *file;
+} psdisk_common;
+
+
+static void psdisk_help(void)
+{
+	printf("\nUsage:");
+	printf("\n\tpsdisk [options]");
+	printf("\n\nOptions");
+	printf("\n%s %-35s %-30s", "   - i ", "<path-to-image-output>", "= Specify image path");
+	printf("\n%s %-35s %-30s", "   - s ", "<mem-size;sector-size>", "= Specify memory's parameters");
+	printf("\n%s %-35s %-30s", "   - p ", "<name;offset;size;type-hex-value>", "= Specify partition's parameters");
+	printf("\n%s %-35s %-30s", "       ", " ", "   types: raw partition = 0x51,");
+	printf("\n%s %-35s %-30s", "       ", " ", "          meterfs partition = 0x75.");
+	printf("\n%s %-35s %-30s", "   - h ", " ", "= Show help");
+	printf("\n\n");
+}
+
+
+static ssize_t psdisk_writeData(unsigned int addr, const void *buff, size_t bufflen)
+{
+	/* TO DO: add option to move partition table at the end of an image ?? */
+	return fwrite(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
+}
+
+
+static int psdisk_createpTable(void)
+{
+	psdisk_common.mem.write = psdisk_writeData;
+
+	/* TODO: call ptable_writePartitions function */
+
+
+	return 0;
+}
+
+
+static int psdisk_parsePartitionOpt(int optind, char *argv[])
+{
+	int i;
+	char *name;
+	ptable_partition_t pHeader;
+
+	const int MAX_PARTITION_ATTR = 4;
+
+	for (i = 0; i < MAX_PARTITION_ATTR; ++i) {
+		if (argv[optind + i] == NULL) {
+			LOG_ERROR("Invalid number of arguments.");
+			return -1;
+		}
+	}
+
+	name = argv[optind];
+	for (i = 0; i < strlen(name); ++i) {
+		if (!isalnum(name[i])) {
+			LOG_ERROR("Invalid partition name - %s.", name);
+			return -1;
+		}
+	}
+
+	if (strlen(name) > sizeof(pHeader.name)) {
+		LOG_ERROR("Invalid partition name - %s.", name);
+		return -1;
+	}
+	else {
+		strncpy((char *)pHeader.name, name, strlen(name));
+	}
+
+	if ((pHeader.offset = strtoul(argv[optind + 1], NULL, 10)) == 0) {
+		LOG_ERROR("Partition %s has invalid offset. %d", pHeader.name);
+		return -1;
+	}
+
+	if ((pHeader.size = strtoul(argv[optind + 2], NULL, 10)) == 0) {
+		LOG_ERROR("Partition %s has invalid size.", pHeader.name);
+		return -1;
+	}
+
+	if ((pHeader.type = strtoul(argv[optind + 3], NULL, 16)) == 0) {
+		LOG_ERROR("Partition %s has invalid type.", pHeader.name);
+		return -1;
+	}
+
+	/* TODO: add pHeader to partitions list */
```

Comment:
TODO: verify if partition is "correct"? Or we would like to verify the whole partition table?